### DISCUSSION TITLE
image preprocessing layers in keras_cv vs keras_hub vs keras ?

### DISCUSSION THREAD

**Discussion by matpalm** [keras-team/keras#20551] [General]
guessing the plan is to move all preprocessing to run as backend agnositic ( using keras.ops ) 
under https://github.com/keras-team/keras/tree/master/keras/src/layers/preprocessing/image_preprocessing
and deprecate the image parts of https://github.com/keras-team/keras-cv/tree/master/keras_cv/src/layers/preprocessing ?

i see the keras_cv is being moved to keras_hub, so what does that mean for 
https://github.com/keras-team/keras-hub/tree/master/keras_hub/src/layers/preprocessing ?

**Comment by fchollet** [comment]
Model-related functionality from KerasCV is being moved to KerasHub. However preprocessing and augmentation layers belong in core Keras (including bbox preprocessing functionality), so we're moving that here. For the time being you can keep using the KerasCV implementations though. Was there a particular layer you were missing?

**Comment by matpalm** [comment]
https://github.com/keras-team/keras-cv/blob/master/keras_cv/src/layers/preprocessing/mosaic.py is the main one; for a lot of datasets ( with yolo like models at least ) i see mosaic makes a BIG difference

since i'm interested in training relatively small models i often have free GPU capacity, so i'm motivated to include augmentations in the training graph directly where possible i.e. not as part of a loading `tf.data` pipeline. i also see with small models that augmentation on CPU can take a reasonable chunk of the wall clock time itself! so i'd definitely love to have them all based on `keras.op`s for training with the jax backend.

**Reply by fchollet** [reply]
That layer only appears to use TF ops that are readily available in Keras 3, so it should be simple to port it. Would you be willing to open a PR?

**Comment by matpalm** [comment]
Yeah sure, I'll give it go! ( Also suspect it might have a bug re: dropping
bboxes occasionally so am keen to review the code deeper )

**Comment by james77777778** [comment]
FYI https://github.com/keras-team/keras/discussions/20050

I previously developed some useful preprocessing/augmentation layers using `keras.ops`.
However, I don't have the bandwidth to port them to Keras.
